src/core/geometry/qgslinestring.cpp

```diff
@@ -2387,3 +2387,166 @@ QgsLineString *QgsLineString::measuredLine( double start, double end ) const
 
   return cloned.release();
 }
+
+QgsLineString *QgsLineString::interpolateM() const
+{
+  if ( !isMeasure() )
+    return nullptr;
+
+  const int totalPoints = numPoints();
+  if ( totalPoints < 2 )
+    return clone();
+
+  const double *xData = mX.constData();
+  const double *yData = mY.constData();
+  const double *mData = mM.constData();
+  const double *zData = is3D() ? mZ.constData() : nullptr;
+
+  QVector< double > xOut( totalPoints );
+  QVector< double > yOut( totalPoints );
+  QVector< double > mOut( totalPoints );
+  QVector< double > zOut( is3D() ? totalPoints : 0 );
+
+  double *xOutData = xOut.data();
+  double *yOutData = yOut.data();
+  double *mOutData = mOut.data();
+  double *zOutData = is3D() ? zOut.data() : nullptr;
+
+  int i = 0;
+  double currentSegmentLength = 0;
+  double lastValidM = std::numeric_limits< double >::quiet_NaN();
+  double prevX = *xData;
+  double prevY = *yData;
+  while ( i < totalPoints )
+  {
+    double thisX = *xData++;
+    double thisY = *yData++;
+    double thisZ = zData ? *zData++ : 0;
+    double thisM = *mData++;
+
+    currentSegmentLength = QgsGeometryUtilsBase::distance2D( prevX, prevY, thisX, thisY );
```

Comment:
Shouldn't we consider 3D distances when the geometry has a Z dimension?

Reply:
Done